### PR TITLE
Sklearn compatibility

### PR DESCRIPTION
**Pull request by PhrankBrochu**
Pull into py3 since it split from there.

Mainly for the modification in random_state behavior in __init__.

Added small test script in /examples/ to test GridSearchCV and Pipeline compatibility. Could be included in the tests eventually (ensure no branch breaks sklearn compatibility)

### CODE REVIEW

File: pyscm/scm.py [hunk 1]
```diff
@@ -110,7 +110,8 @@ def fit(self, X, y, iteration_callback=None, **fit_params):
 
             # TODO: Support user specified tiebreaker
             logging.debug("Tiebreaking. Found %d optimal rules" % len(opti_feat_idx))
-            keep_idx = self.random_state.randint(0, len(opti_feat_idx))
+            random_state = check_random_state(self.random_state)
```
Comment (aldro61) [inline pyscm/scm.py]: @PhrankBrochu Once this is done, I'll merge you PR. Also, next time, I think you can just use `git pull origin sklearn-compatibility` from within your branch. You should also keep it up to date with master in the same way: `git pull origin master` and resolve any conflicts manually.

**Comment by plpla** [comment]
Currently, the estimator does not support multi-class and therefore is not fully compatible with sklearn (from what I understand). I propose to implement a one-vs-all approach for multi-class problems where class with labels other than 0 are changed to 1.

File: pyscm/scm.py [hunk 2]
```diff
@@ -255,7 +255,7 @@ class SetCoveringMachineClassifier(BaseSetCoveringMachine):
         The random state.
 
     """
-    def __init__(self, p=1.0, model_type="conjunction", max_rules=10, random_state=None):
+    def __init__(self, p=1.0, model_type=str("conjunction"), max_rules=10, random_state=None):
```
Comment (aldro61) [inline pyscm/scm.py]: Why did you add str() around the string? It is already a string.

File: pyscm/tests/utility_tests.py [hunk 3]
```diff
@@ -152,4 +153,46 @@ def automatic_testing(self):
                         P_bar = 1.0 * (~rule_classifications[y == 1]).sum()
                         g_rule_utilities.append(N - p * P_bar)
 
-                    np.testing.assert_almost_equal(actual=solver_best_utility, desired=max(max(le_rule_utilities), max(g_rule_utilities)))
+                    np.testing.assert_almost_equal(actual=solver_best_utility, desired=max(max(le_rule_utilities),
+                                                                                           max(g_rule_utilities)))
+
+    def sklearn_compatibility_test(self):
```
Comment (aldro61) [inline pyscm/tests/utility_tests.py]: Can you please put these tests in a new test file called test_sklearn_compatibility.py. Use the same model as the utility_tests.py file.

Comment (plpla) [inline pyscm/tests/utility_tests.py]: Will do.

File: pyscm/scm.py (hunk 2 above)
Comment (plpla) [inline pyscm/scm.py]: It was seen as unicode by sklearn test suit. I did not found another work-around. Any idea?

**Comment by plpla** [comment]
There is still a problem with the `GridSearchCV` function when creating more than 2 jobs. I am investigating.

File: pyscm/scm.py [hunk 4]
```diff
@@ -110,7 +126,8 @@ def fit(self, X, y, iteration_callback=None, **fit_params):
 
             # TODO: Support user specified tiebreaker
             logging.debug("Tiebreaking. Found %d optimal rules" % len(opti_feat_idx))
-            keep_idx = self.random_state.randint(0, len(opti_feat_idx))
+            random_state = check_random_state(self.random_state)
```
Comment (aldro61) [inline pyscm/scm.py]: @plpla This is still a bug. The random state should be checked at the beginning of the fit function. If you check at each tiebreak, you will always generate the same number (in the case where the random_state is an int).

**Comment by aldro61** [comment]
@plpla Since you removed all checks from the constructor, I think that we dont need the get_params and set_params functions to be defined in the BaseSetCoveringMachine class. We should be good with the default ones inherited from sklearn.
